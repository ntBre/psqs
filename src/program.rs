use symm::Atom;

use crate::geom::Geom;

pub mod molpro;
pub mod mopac;

#[derive(Clone, Debug, Default, PartialEq)]
pub struct ProgramResult {
    pub energy: f64,
    pub cart_geom: Option<Vec<Atom>>,
}

#[derive(Debug, PartialEq, Eq)]
pub enum ProgramError {
    FileNotFound(String),
    ErrorInOutput(String),
    EnergyNotFound(String),
    EnergyParseError(String),
}

impl ProgramError {
    /// Returns `true` if the program error is [`ErrorInOutput`].
    ///
    /// [`ErrorInOutput`]: ProgramError::ErrorInOutput
    #[must_use]
    pub fn is_error_in_output(&self) -> bool {
        matches!(self, Self::ErrorInOutput(..))
    }
}

#[derive(Debug, PartialEq, Eq, Copy, Clone)]
pub enum Procedure {
    Opt,
    Freq,
    SinglePt,
}

#[derive(Clone, Debug)]
pub struct Template {
    header: String,
}

impl Template {
    pub fn from(s: &str) -> Self {
        Self {
            header: s.to_string(),
        }
    }
}

/// A trait for describing programs runnable on a [crate::queue::Queue]
pub trait Program {
    /// returns the file associated with the program's input. it should not
    /// include an extension
    fn filename(&self) -> String;

    /// set `filename`
    fn set_filename(&mut self, filename: &str);

    /// the template for writing input files
    fn template(&self) -> &Template;

    /// the file extension for the input file
    fn extension(&self) -> String;

    /// molecular charge
    fn charge(&self) -> isize;

    /// write the input file to the name returned by `filename`
    fn write_input(&mut self, proc: Procedure);

    /// read the output file found by replacing `self.extension()` with `.out`
    fn read_output(&self) -> Result<ProgramResult, ProgramError>;

    /// Return all the filenames associated with the Program for deletion when
    /// it finishes
    fn associated_files(&self) -> Vec<String>;

    fn new(
        filename: String,
        template: Template,
        charge: isize,
        geom: Geom,
    ) -> Self;

    /// Build the jobs described by `moles` in memory, but don't write any of
    /// their files yet
    fn build_jobs(
        moles: &Vec<Geom>,
        dir: &'static str,
        start_index: usize,
        coeff: f64,
        job_num: usize,
        charge: isize,
        tmpl: Template,
    ) -> Vec<Job<Self>>
    where
        Self: std::marker::Sized,
    {
        let mut count: usize = start_index;
        let mut job_num = job_num;
        let mut jobs = Vec::new();
        for mol in moles {
            let filename = format!("{dir}/job.{:08}", job_num);
            job_num += 1;
            let mut job = Job::new(
                Self::new(filename, tmpl.clone(), charge, mol.clone()),
                count,
            );
            job.coeff = coeff;
            jobs.push(job);
            count += 1;
        }
        jobs
    }
}

#[derive(Debug, Clone)]
pub struct Job<P: Program> {
    pub program: P,
    pub pbs_file: String,
    pub job_id: String,

    /// the index in the output array to store the result
    pub index: usize,

    /// the coefficient to multiply by when storing the result
    pub coeff: f64,
}

impl<P: Program> Job<P> {
    pub fn new(program: P, index: usize) -> Self {
        Self {
            program,
            pbs_file: String::new(),
            job_id: String::new(),
            index,
            coeff: 1.0,
        }
    }
}
