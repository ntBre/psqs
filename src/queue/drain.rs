use core::time;
use std::{
    collections::{HashMap, HashSet},
    fmt::Display,
    thread,
    time::Duration,
};

use crate::{
    geom::Geom,
    program::{Job, Procedure, Program, ProgramError, ProgramResult},
    queue::drain::dump::Dump,
};

use super::{Queue, DEBUG};

macro_rules! time {
    ($msg:expr, $body:block) => {
        let now = std::time::Instant::now();
        $body;
        eprintln!(
            "finished {} after {:.1} s",
            $msg,
            now.elapsed().as_millis() as f64 / 1000.0
        );
    };
}

mod dump;

#[derive(Default)]
struct Time {
    writing: Duration,
    reading: Duration,
    sleeping: Duration,
    removing: Duration,
}

impl Display for Time {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{:.1} s reading ok, {:.1} s writing, {:.1} s sleeping, \
	     {:.1} s removing",
            self.reading.as_millis() as f64 / 1000.0,
            self.writing.as_millis() as f64 / 1000.0,
            self.sleeping.as_millis() as f64 / 1000.0,
            self.removing.as_millis() as f64 / 1000.0,
        )
    }
}

/// a histogram covering the range [min, max) with `N` bins
struct Histogram<const N: usize> {
    min: f64,
    cur_min: f64,
    cur_max: f64,
    total: f64,
    denom: f64,
    data: [usize; N],
}

impl<const N: usize> Histogram<N> {
    fn new(min: f64, max: f64) -> Self {
        Self {
            min,
            denom: max - min,
            data: [0; N],
            cur_min: 0.0,
            cur_max: 0.0,
            total: 0.0,
        }
    }

    /// insert `val` into the appropriate bin in `self` and add it to the total.
    /// if `val` is greater than `self.max`, don't perform the insert but add it
    /// to the other statistics
    fn insert(&mut self, val: f64) {
        let idx = N as f64 * (val - self.min) / self.denom;
        if let Some(elt) = self.data.get_mut(idx.floor() as usize) {
            *elt += 1;
        }
        if val > self.cur_max {
            self.cur_max = val;
        }
        if val < self.cur_min {
            self.cur_min = val;
        }
        self.total += val;
    }

    /// return the count of elements in `self`
    fn count(&self) -> usize {
        self.data.iter().sum()
    }

    /// return the average of `self`
    fn average(&self) -> f64 {
        self.total / self.count() as f64
    }
}

impl<const N: usize> Display for Histogram<N> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let bin_width = self.denom / N as f64;
        for (i, v) in self.data.iter().enumerate() {
            if *v > 0 {
                writeln!(f, "{:5.2}{:5}", i as f64 * bin_width, v)?;
            }
        }
        Ok(())
    }
}

pub(crate) trait Drain {
    type Item;

    fn procedure(&self) -> Procedure;

    fn set_result<P: Program>(
        &self,
        dst: &mut [Self::Item],
        job: &mut Job<P>,
        res: ProgramResult,
    );

    fn drain<P: Program + Clone, Q: Queue<P> + ?Sized>(
        &self,
        dir: &str,
        queue: &Q,
        jobs: &mut [Job<P>],
        dst: &mut [Self::Item],
    ) -> Result<(), ProgramError> {
        let mut chunk_num: usize = 0;
        let mut cur_jobs = Vec::new();
        let mut slurm_jobs = HashMap::new();
        let mut remaining = jobs.len();

        let dump = Dump::new();
        let mut time = Time::default();

        // histogram for tracking job times
        let mut job_time = Histogram::<100>::new(0.0, 10.0);

        let mut qstat = HashSet::<String>::new();
        let mut chunks = jobs.chunks_mut(queue.chunk_size());
        let mut out_of_jobs = false;
        let mut to_remove = Vec::new();
        loop {
            let loop_time = std::time::Instant::now();
            // build more jobs if there is room
            while !out_of_jobs && cur_jobs.len() < queue.job_limit() {
                match chunks.next() {
                    Some(jobs) => {
                        let now = std::time::Instant::now();
                        queue.build_chunk(
                            dir,
                            jobs,
                            chunk_num,
                            &mut slurm_jobs,
                            self.procedure(),
                        );
                        let job_id = jobs[0].job_id.clone();
                        qstat.insert(job_id);
                        let elapsed = now.elapsed();
                        if DEBUG {
                            eprintln!(
                                "submitted chunk {} after {:.1} s",
                                chunk_num,
                                elapsed.as_millis() as f64 / 1000.0
                            );
                        }
                        time.writing += elapsed;
                        chunk_num += 1;
                        cur_jobs.extend(jobs);
                    }
                    None => {
                        out_of_jobs = true;
                        break;
                    }
                }
            }
            // collect output
            let mut finished = 0;
            to_remove.clear();
            let now = std::time::Instant::now();
            let outfiles: Vec<_> =
                cur_jobs.iter().map(|job| job.program.filename()).collect();
            use rayon::prelude::*;
            let results: Vec<_> =
                outfiles.par_iter().map(|out| P::read_output(out)).collect();
            time.reading += now.elapsed();
            for (i, (job, res)) in cur_jobs.iter_mut().zip(results).enumerate()
            {
                match res {
                    Ok(res) => {
                        to_remove.push(i);
                        job_time.insert(res.time);
                        self.set_result(dst, *job, res);
                        for f in job.program.associated_files() {
                            dump.send(f);
                        }
                        finished += 1;
                        remaining -= 1;
                        let job_name = job.pbs_file.as_str();
                        let mut count = match slurm_jobs.get_mut(job_name) {
                            Some(n) => *n,
                            None => {
                                eprintln!(
                                    "failed to find {} in slurm_jobs",
                                    job_name
                                );
                                1
                            }
                        };
                        count -= 1;
                        if count == 0 {
                            // delete the submit script and output file
                            dump.send(job_name.to_string());
                            dump.send(format!("{}.out", job_name));
                        }
                    }
                    Err(e) => {
                        if e.is_error_in_output() {
                            dump.shutdown();
                            return Err(e);
                        }
                        queue.drain_err_case(
                            e,
                            &mut qstat,
                            &mut slurm_jobs,
                            job,
                        );
                    }
                }
            }
            // have to remove the highest index first so sort and reverse
            let r = std::time::Instant::now();
            to_remove.sort();
            to_remove.reverse();
            for i in &to_remove {
                cur_jobs.swap_remove(*i);
            }
            time.removing += r.elapsed();
            if DEBUG {
                eprintln!(
                    "finished {} jobs in {:.1} s",
                    finished,
                    loop_time.elapsed().as_millis() as f64 / 1000.0
                );
            }
            if cur_jobs.is_empty() && out_of_jobs {
                dump.shutdown();
                eprintln!("{}", time);
                eprintln!("total job time: {:.2} s", job_time.total);
                eprintln!("max job time: {:.2} s", job_time.cur_max);
                eprintln!("min job time: {:.2} s", job_time.cur_min);
                eprintln!("avg job time: {:.2} s", job_time.average());
                eprint!("histogram:\n{}", job_time);
                return Ok(());
            }
            if finished == 0 {
                eprintln!("{} jobs remaining", remaining);
                qstat = queue.status();
                let d = time::Duration::from_secs(queue.sleep_int() as u64);
                time.sleeping += d;
                thread::sleep(d);
            }
        }
    }
}

pub(crate) struct Opt;

impl Drain for Opt {
    type Item = Geom;

    fn procedure(&self) -> Procedure {
        Procedure::Opt
    }

    fn set_result<P: Program>(
        &self,
        dst: &mut [Self::Item],
        job: &mut Job<P>,
        res: ProgramResult,
    ) {
        dst[job.index] = Geom::Xyz(res.cart_geom.unwrap());
    }
}

pub(crate) struct Single;

impl Drain for Single {
    type Item = f64;

    fn procedure(&self) -> Procedure {
        Procedure::SinglePt
    }

    fn set_result<P: Program>(
        &self,
        dst: &mut [Self::Item],
        job: &mut Job<P>,
        res: ProgramResult,
    ) {
        dst[job.index] += job.coeff * res.energy;
    }
}

pub(crate) struct Both;

impl Drain for Both {
    type Item = ProgramResult;

    fn procedure(&self) -> Procedure {
        Procedure::Opt
    }

    fn set_result<P: Program>(
        &self,
        dst: &mut [Self::Item],
        job: &mut Job<P>,
        res: ProgramResult,
    ) {
        dst[job.index] = res;
    }
}
