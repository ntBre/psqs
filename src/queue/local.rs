use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::fs::File;
use std::io::Write;

use crate::program::dftbplus::DFTBPlus;
use crate::program::molpro::Molpro;
use crate::program::{mopac::Mopac, Program};
use crate::queue::Queue;

use super::{SubQueue, Submit};

/// Minimal implementation for testing MOPAC locally
#[derive(Debug)]
pub struct Local {
    pub dir: String,
    pub chunk_size: usize,
    pub mopac: String,
}

impl Default for Local {
    fn default() -> Self {
        Self {
            dir: ".".to_string(),
            chunk_size: 128,
            mopac: "/opt/mopac/mopac".to_owned(),
        }
    }
}

impl Local {
    pub fn new(
        chunk_size: usize,
        _job_limit: usize,
        _sleep_int: usize,
        dir: &'static str,
        _no_del: bool,
        _template: Option<String>,
    ) -> Self {
        Self {
            dir: dir.to_string(),
            chunk_size,
            mopac: "/opt/mopac/mopac".to_string(),
        }
    }
}

impl Submit<Molpro> for Local {}

impl Queue<Molpro> for Local {
    fn default_submit_script(&self) -> String {
        todo!()
    }

    fn write_submit_script(&self, _infiles: &[String], _filename: &str) {
        todo!()
    }
}

impl Submit<Mopac> for Local {}

impl Queue<Mopac> for Local {
    fn write_submit_script(&self, infiles: &[String], filename: &str) {
        use std::fmt::Write;
        let mut body = String::from("export LD_LIBRARY_PATH=/opt/mopac/\n");
        for f in infiles {
            writeln!(body, "{} {f}.mop &> {filename}.out", self.mopac).unwrap();
            writeln!(body, "cat {f}.mop {f}.out >> {filename}.out").unwrap();
            writeln!(body, "echo \"================\" >> {filename}.out")
                .unwrap();
        }
        writeln!(body, "date +%s >> {filename}.out").unwrap();
        let mut file = File::create(filename).unwrap_or_else(|_| {
            panic!("failed to create submit script `{filename}`")
        });
        write!(file, "{body}").expect("failed to write submit script");
    }

    fn default_submit_script(&self) -> String {
        todo!()
    }
}

impl Submit<DFTBPlus> for Local {}

impl Queue<DFTBPlus> for Local {
    fn default_submit_script(&self) -> String {
        todo!()
    }

    fn write_submit_script(&self, infiles: &[String], filename: &str) {
        use std::fmt::Write;
        let mut body = String::new();
        // assume f is a directory name, not a real file
        let c = std::env::var("DFTB_PATH").unwrap_or("/opt/dftb+/dftb+".into());
        for f in infiles {
            writeln!(body, "(cd {f} && {c} > out)").unwrap();
        }
        writeln!(body, "date +%s >> {filename}.out").unwrap();
        let mut file = File::create(filename).unwrap_or_else(|_| {
            panic!("failed to create submit script `{filename}`")
        });
        write!(file, "{body}").expect("failed to write submit script");
    }
}

impl<P: Program + Clone + Serialize + for<'a> Deserialize<'a>> SubQueue<P>
    for Local
{
    fn submit_command(&self) -> &str {
        "bash"
    }

    fn chunk_size(&self) -> usize {
        self.chunk_size
    }

    fn job_limit(&self) -> usize {
        1600
    }

    fn sleep_int(&self) -> usize {
        1
    }

    const SCRIPT_EXT: &'static str = "slurm";

    fn dir(&self) -> &str {
        &self.dir
    }

    fn stat_cmd(&self) -> String {
        todo!()
    }

    fn status(&self) -> HashSet<String> {
        for dir in ["opt", "pts", "freqs"] {
            let d = std::fs::read_dir(dir).unwrap();
            for f in d {
                eprintln!("contents of {:?}", f.as_ref().unwrap());
                eprintln!(
                    "{}",
                    std::fs::read_to_string(f.unwrap().path()).unwrap()
                );
                eprintln!("================");
            }
        }
        panic!("no status available for Local queue");
    }

    fn no_del(&self) -> bool {
        false
    }
}
