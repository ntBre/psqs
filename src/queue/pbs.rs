use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::{collections::HashSet, process::Command};

use serde::{Deserialize, Serialize};

use crate::program::molpro::Molpro;
use crate::program::mopac::Mopac;
use crate::program::Program;
use crate::queue::Queue;

use super::{SubQueue, Submit};

/// Pbs is a type for holding the information for submitting a pbs job.
/// `filename` is the name of the Pbs submission script
#[derive(Debug)]
pub struct Pbs {
    chunk_size: usize,
    job_limit: usize,
    sleep_int: usize,
    dir: &'static str,
}

impl Pbs {
    pub fn new(
        chunk_size: usize,
        job_limit: usize,
        sleep_int: usize,
        dir: &'static str,
    ) -> Self {
        Self {
            chunk_size,
            job_limit,
            sleep_int,
            dir,
        }
    }
}

impl Default for Pbs {
    fn default() -> Self {
        Self {
            chunk_size: 128,
            job_limit: 1600,
            sleep_int: 5,
            dir: "inp",
        }
    }
}

impl Submit<Mopac> for Pbs
where
    Mopac: Serialize + for<'a> Deserialize<'a>,
{
    /// submit `filename` to the queue and return the jobid
    fn submit(&self, filename: &str) -> String {
        match Command::new(<Self as SubQueue<Mopac>>::submit_command(self))
            .arg("-f")
            .arg(filename)
            .output()
        {
            Ok(s) => {
                let raw =
                    std::str::from_utf8(&s.stdout).unwrap().trim().to_string();
                return raw
                    .split_whitespace()
                    .last()
                    .unwrap_or("no jobid")
                    .to_string();
            }
            Err(e) => panic!("{e:?}"),
        };
    }
}

// Molpro 2022 submit script requires submission from the current directory, so
// we have to override the default impl
impl Submit<Molpro> for Pbs
where
    Molpro: Serialize + for<'a> Deserialize<'a>,
{
    fn submit(&self, filename: &str) -> String {
        let path = Path::new(filename);
        let dir = path.parent().unwrap();
        let base = path.file_name().unwrap();
        let mut cmd =
            Command::new(<Self as SubQueue<Molpro>>::submit_command(self));
        let cmd = cmd.arg(base).current_dir(dir);
        match cmd.output() {
            Ok(s) => {
                let raw =
                    std::str::from_utf8(&s.stdout).unwrap().trim().to_string();
                return raw.split_whitespace().last().unwrap().to_string();
            }
            Err(e) => panic!("{e:?}"),
        };
    }
}

impl Queue<Molpro> for Pbs
where
    Molpro: Serialize + for<'a> Deserialize<'a>,
{
    fn write_submit_script(&self, infiles: &[String], filename: &str) {
        let path = Path::new(filename);
        let basename = path.file_name().unwrap();
        let mut body = format!(
            "#!/bin/sh
#PBS -N {basename:?}
#PBS -S /bin/bash
#PBS -j oe
#PBS -o {basename:?}.out
#PBS -W umask=022
#PBS -l walltime=9999:00:00
#PBS -l ncpus=1
#PBS -l mem=8gb
#PBS -q workq

module load openpbs molpro

export WORKDIR=$PBS_O_WORKDIR
export TMPDIR=/tmp/$USER/$PBS_JOBID
cd $WORKDIR
mkdir -p $TMPDIR

"
        );
        {
            use std::fmt::Write;
            for f in infiles {
                let basename = Path::new(f).file_name().unwrap();
                writeln!(body, "molpro -t 1 --no-xml-output {basename:?}.inp")
                    .unwrap();
            }
            writeln!(body, "rm -rf $TMPDIR").unwrap();
        }
        let mut file = match File::create(filename) {
            Ok(f) => f,
            Err(_) => {
                panic!("write_submit_script: failed to create {filename}");
            }
        };
        write!(file, "{}", body).unwrap_or_else(|_| {
            panic!("failed to write molpro input file: {}", filename)
        });
    }
}

impl Queue<Mopac> for Pbs {
    fn write_submit_script(&self, infiles: &[String], filename: &str) {
        let path = Path::new(filename);
        let basename = path.file_name().unwrap();
        let mut body = format!(
            "#!/bin/sh
#PBS -N {basename:?}
#PBS -S /bin/bash
#PBS -j oe
#PBS -o {filename}.out
#PBS -W umask=022
#PBS -l walltime=9999:00:00
#PBS -l ncpus=1
#PBS -l mem=1gb
#PBS -q workq

module load openpbs

export WORKDIR=$PBS_O_WORKDIR
cd $WORKDIR

",
        );
        for f in infiles {
            body.push_str(&format!(
                "/ddn/home1/r2518/Packages/mopac/build/mopac {f}.mop\n"
            ));
        }
        let mut file = match File::create(filename) {
            Ok(f) => f,
            Err(_) => {
                eprintln!("write_submit_script: failed to create {filename}");
                std::process::exit(1);
            }
        };
        write!(file, "{}", body).expect("failed to write params file");
    }
}

impl<P: Program + Clone + Serialize + for<'a> Deserialize<'a>> SubQueue<P>
    for Pbs
{
    fn submit_command(&self) -> &str {
        "qsub"
    }

    fn chunk_size(&self) -> usize {
        self.chunk_size
    }

    fn job_limit(&self) -> usize {
        self.job_limit
    }

    fn sleep_int(&self) -> usize {
        self.sleep_int
    }

    const SCRIPT_EXT: &'static str = "pbs";

    fn dir(&self) -> &str {
        self.dir
    }

    /// run `qstat -u $USER`. form of the output is:
    ///
    /// maple:
    ///                                                             Req'd  Req'd   Elap
    /// Job ID          Username Queue    Jobname    SessID NDS TSK Memory Time  S Time
    /// --------------- -------- -------- ---------- ------ --- --- ------ ----- - -----
    /// 819446          user     queue    C6HNpts      5085   1   1    8gb 26784 R 00:00
    fn stat_cmd(&self) -> String {
        let user = std::env::vars()
            .find(|x| x.0 == "USER")
            .expect("couldn't find $USER env var");
        let status = match Command::new("qstat").args(["-u", &user.1]).output()
        {
            Ok(status) => status,
            Err(e) => panic!("failed to run squeue with {}", e),
        };
        String::from_utf8(status.stdout)
            .expect("failed to convert squeue output to String")
    }

    fn status(&self) -> HashSet<String> {
        let mut ret = HashSet::new();
        let lines = <Pbs as SubQueue<P>>::stat_cmd(self);
        // skip to end of header
        let lines = lines.lines().skip_while(|l| !l.contains("-----------"));
        for line in lines {
            let fields: Vec<_> = line.split_whitespace().collect();
            assert!(fields.len() == 11);
            ret.insert(fields[0].to_string());
        }
        ret
    }
}
