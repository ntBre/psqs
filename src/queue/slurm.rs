use std::collections::HashSet;
use std::fs::File;
use std::io::Write;

use serde::{Deserialize, Serialize};

use crate::program::mopac::Mopac;
use crate::program::Program;
use crate::program::{dftbplus::DFTBPlus, molpro::Molpro};
use crate::queue::Queue;

use super::{SubQueue, Submit};

/// Slurm is a type for holding the information for submitting a slurm job.
/// `filename` is the name of the Slurm submission script
#[derive(Debug)]
pub struct Slurm {
    chunk_size: usize,
    job_limit: usize,
    sleep_int: usize,
    dir: &'static str,
    no_del: bool,
    template: Option<String>,
}

impl Slurm {
    pub fn new(
        chunk_size: usize,
        job_limit: usize,
        sleep_int: usize,
        dir: &'static str,
        no_del: bool,
        template: Option<String>,
    ) -> Self {
        Self {
            chunk_size,
            job_limit,
            sleep_int,
            dir,
            no_del,
            template,
        }
    }
}

impl<P: Program + Clone + Serialize + for<'a> Deserialize<'a>> Submit<P>
    for Slurm
{
}

impl Queue<Molpro> for Slurm {
    fn write_submit_script(
        &self,
        infiles: impl IntoIterator<Item = String>,
        filename: &str,
    ) {
        let mut body = self
            .template
            .clone()
            .unwrap_or_else(|| {
                <Self as Queue<Molpro>>::default_submit_script(self)
            })
            .replace("{{.filename}}", filename);
        for f in infiles {
            body.push_str(&format!("$MOLPRO_CMD {f}.inp\n"));
        }
        let mut file = match File::create(filename) {
            Ok(f) => f,
            Err(_) => {
                eprintln!("write_submit_script: failed to create {filename}");
                std::process::exit(1);
            }
        };
        write!(file, "{body}").unwrap_or_else(|_| {
            panic!("failed to write molpro input file: {filename}")
        });
    }

    fn default_submit_script(&self) -> String {
        "#!/bin/bash
#SBATCH --job-name={{.filename}}
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=1
#SBATCH -o {{.filename}}.out
#SBATCH --no-requeue
#SBATCH --mem=8gb

MOLPRO_CMD=\"/home/qc/bin/molpro2020.sh 1 1\"
"
        .to_owned()
    }
}

impl Queue<Mopac> for Slurm {
    fn write_submit_script(
        &self,
        infiles: impl IntoIterator<Item = String>,
        filename: &str,
    ) {
        let mut body = self
            .template
            .clone()
            .unwrap_or_else(|| {
                <Self as Queue<Mopac>>::default_submit_script(self)
            })
            .replace("{{.filename}}", filename);
        for f in infiles {
            body.push_str(&format!(
                "/home/qc/mopac2016/MOPAC2016.exe {f}.mop\n"
            ));
        }
        let mut file = match File::create(filename) {
            Ok(f) => f,
            Err(_) => {
                eprintln!("write_submit_script: failed to create {filename}");
                std::process::exit(1);
            }
        };
        write!(file, "{body}").expect("failed to write params file");
    }

    fn default_submit_script(&self) -> String {
        "#!/bin/bash
#SBATCH --job-name=semp
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=1
#SBATCH -o {{.filename}}.out
#SBATCH --no-requeue
#SBATCH --mem=1gb
export LD_LIBRARY_PATH=/home/qc/mopac2016/
echo $SLURM_JOB_ID
date
hostname\n"
            .to_owned()
    }
}

impl Queue<DFTBPlus> for Slurm {
    fn default_submit_script(&self) -> String {
        todo!()
    }

    fn write_submit_script(
        &self,
        _infiles: impl IntoIterator<Item = String>,
        _filename: &str,
    ) {
        todo!()
    }
}

impl<P> SubQueue<P> for Slurm
where
    P: Program + Clone + Serialize + for<'a> Deserialize<'a>,
{
    fn submit_command(&self) -> &str {
        "sbatch"
    }

    fn chunk_size(&self) -> usize {
        self.chunk_size
    }

    fn job_limit(&self) -> usize {
        self.job_limit
    }

    fn sleep_int(&self) -> usize {
        self.sleep_int
    }

    const SCRIPT_EXT: &'static str = "slurm";

    fn dir(&self) -> &str {
        self.dir
    }

    /// run `squeue -u $USER`. form of the output is:
    ///
    ///    JOBID PARTITION   NAME     USER ST        TIME  NODES NODELIST(REASON)
    /// 30627992   compute  c3oh-   mdavis  R 46-17:12:23      1 node2
    fn stat_cmd(&self) -> String {
        let user = std::env::var("USER").expect("couldn't find $USER env var");
        let status = match std::process::Command::new("squeue")
            .args(["-u", &user])
            .output()
        {
            Ok(status) => status,
            Err(e) => panic!("failed to run squeue with {e}"),
        };
        String::from_utf8(status.stdout)
            .expect("failed to convert squeue output to String")
    }

    fn status(&self) -> HashSet<String> {
        let mut ret = HashSet::new();
        // wut?
        let lines = <Slurm as SubQueue<P>>::stat_cmd(self);
        let lines = lines.lines();
        for line in lines {
            if !line.contains("JOBID") {
                let fields: Vec<_> = line.split_whitespace().collect();
                assert!(fields.len() == 8);
                // exclude completing jobs to combat stuck completing bug
                if fields[4] != "CG" {
                    ret.insert(fields[0].to_string());
                }
            }
        }
        ret
    }

    fn no_del(&self) -> bool {
        self.no_del
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::*;

    fn slurm() -> Slurm {
        Slurm {
            chunk_size: 1,
            job_limit: 1,
            sleep_int: 1,
            dir: "/tmp",
            no_del: false,
            template: None,
        }
    }

    macro_rules! make_tests {
        ($($name:ident, $queue:expr => $q:ty, $p:ty$(,)*)*) => {
            $(
            #[test]
            fn $name() {
                let tmp = tempfile::NamedTempFile::new().unwrap();
                <$q as Queue<$p>>::write_submit_script(
                    $queue,
                    ["opt0.inp", "opt1.inp", "opt2.inp", "opt3.inp"].map(|s| s.into()),
                    tmp.path().to_str().unwrap(),
                );
                let got = std::fs::read_to_string(tmp).unwrap();
                let got: Vec<&str> = got.lines().filter(|l|
                    !(l.starts_with("#SBATCH --job-name")
                        || l.starts_with("#SBATCH -o"))).collect();
                let got = got.join("\n");
                assert_snapshot!(got);
            }
            )*
        }
    }

    make_tests! {
        mopac_slurm, &slurm() => Slurm, Mopac,
        molpro_slurm, &slurm() => Slurm, Molpro,
        // cfour_slurm, &slurm() => Slurm, Cfour,
        // dftb_slurm, &slurm() => Slurm, DFTBPlus,
    }
}
