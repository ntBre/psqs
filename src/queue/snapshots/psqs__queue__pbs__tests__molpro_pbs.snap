---
source: src/queue/pbs.rs
expression: got
---
#!/bin/sh
#PBS -S /bin/bash
#PBS -j oe
#PBS -W umask=022
#PBS -l walltime=1000:00:00
#PBS -l ncpus=1
#PBS -l mem=8gb
#PBS -q workq

module load openpbs molpro

export WORKDIR=$PBS_O_WORKDIR
export TMPDIR=/tmp/$USER/$PBS_JOBID
cd $WORKDIR
mkdir -p $TMPDIR
trap 'rm -rf $TMPDIR' EXIT
molpro -t $NCPUS --no-xml-output opt0.inp.inp
molpro -t $NCPUS --no-xml-output opt1.inp.inp
molpro -t $NCPUS --no-xml-output opt2.inp.inp
molpro -t $NCPUS --no-xml-output opt3.inp.inp
